controllers/node_label_module_version_reconciler.go

```diff
@@ -0,0 +1,257 @@
+)
+
+type reconcileLabelsResult struct {
+	labelsToAdd    map[string]string
+	labelsToDelete []string
+	requeue        bool
+}
+
+type NodeLabelModuleVersionReconciler struct {
+	client    client.Client
+	helperAPI nodeLabelModuleVersionHelperAPI
+}
+
+func NewNodeLabelModuleVersionReconciler(client client.Client) *NodeLabelModuleVersionReconciler {
+	return &NodeLabelModuleVersionReconciler{
+		client:    client,
+		helperAPI: newNodeLabelModuleVersionHelper(client),
+	}
+}
+
+func (nlmvr *NodeLabelModuleVersionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	node := v1.Node{}
+
+	if err := nlmvr.client.Get(ctx, types.NamespacedName{Name: req.Name}, &node); err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not get node: %v", err)
+	}
+
+	logger := log.FromContext(ctx).WithValues("node name", node.Name)
+
+	// get node's label data per module
+	modulesVersionLabels := nlmvr.helperAPI.getLabelsPerModules(node.Labels)
+
+	// get module loader pods on the node
+	moduleLoaderPods, err := nlmvr.helperAPI.getModuleLoaderPods(ctx, node.Name)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could get module loader pods for the node %s: %v", node.Name, err)
+	}
+
+	// get the labels to add and delete for this specific node
+	reconLabelsRes := nlmvr.helperAPI.reconcileLabels(modulesVersionLabels, moduleLoaderPods)
+
+	logLabelsUpdateData(logger, reconLabelsRes)
+
+	// update the node with the labels
+	err = nlmvr.helperAPI.updateNodeLabels(ctx, node.Name, reconLabelsRes)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not label node %s with added/deleted labels: %v", node.Name, err)
+	}
+
+	return ctrl.Result{Requeue: reconLabelsRes.requeue}, nil
+}
+
+//go:generate mockgen -source=node_label_module_version_reconciler.go -package=controllers -destination=mock_node_label_module_version_reconciler.go nodeLabelModuleVersionHelperAPI
+
+type nodeLabelModuleVersionHelperAPI interface {
+	getLabelsPerModules(nodeLabels map[string]string) map[string]*modulesVersionLabels
+	getModuleLoaderPods(ctx context.Context, nodeName string) ([]v1.Pod, error)
+	reconcileLabels(modulesLabels map[string]*modulesVersionLabels, moduleLoaderPods []v1.Pod) *reconcileLabelsResult
+	updateNodeLabels(ctx context.Context, nodeName string, reconLabelsRes *reconcileLabelsResult) error
+}
+
+type nodeLabelModuleVersionHelper struct {
+	client client.Client
+}
+
+func newNodeLabelModuleVersionHelper(client client.Client) nodeLabelModuleVersionHelperAPI {
+	return &nodeLabelModuleVersionHelper{client: client}
+}
+
+func (nlmvha *nodeLabelModuleVersionHelper) getLabelsPerModules(nodeLabels map[string]string) map[string]*modulesVersionLabels {
+	labelsPerModule := map[string]*modulesVersionLabels{}
+	for key, value := range nodeLabels {
+		if utils.IsVersionLabel(key) {
+			namespace, name, err := utils.GetNamespaceNameFromVersionLabel(key)
+			if err != nil {
+				continue
```

Comment:
Why are we skipping here? Why don't we return an error or at least a warning log?